ttnn/cpp/ttnn/operations/ccl/shared_with_host/sharded_tensor_addr_gen.hpp

```diff
@@ -0,0 +1,449 @@
+    worker_to_noc_lookup_t worker_to_noc_lookup;
+    DEVICE_SHARD_SPEC_T tensor_shard_spec;
+    uint32_t page_size;
+    uint32_t bank_base_address;
+
+   public:
+    constexpr WidthShardedAddressGenerator(worker_to_noc_lookup_t lookup, DEVICE_SHARD_SPEC_T const& tensor_shard_spec, uint32_t page_size, uint32_t base_address) : worker_to_noc_lookup(lookup), tensor_shard_spec(tensor_shard_spec), page_size(page_size), bank_base_address(base_address) {}
+
+    test_shard_location_t get_page_location(std::size_t global_page_id) const {
+        // With width sharding, the tensor is fractured along width, but can be mapped onto a 2D grid, in such a case
+        // the logical tensor is still fractured only along 1 dimension but the placement/allocation snakes through the
+        // grid.
+        std::size_t global_pages_per_row_logical = tensor_shard_spec.get_pages_per_tensor_x();
+
+        std::size_t page_global_outer_dim = global_page_id / global_pages_per_row_logical;
+        std::size_t page_global_inner_dim = global_page_id - (page_global_outer_dim * global_pages_per_row_logical);
+
+        // might be able to save on some divides here if we can multiply out the pages per shard_x and shards per row
+        // ... think about it
+        // likewise maybe we can also take it a step further and do the same sort of thing above too to get away with a
+        // single divide for the full function
+        std::size_t global_shard_index = page_global_inner_dim / tensor_shard_spec.get_pages_per_shard_x();
+
+        auto [shard_grid_inner_dim_index, shard_grid_outer_dim_index] = flat_index_to_2d<std::size_t>(global_shard_index, tensor_shard_spec.get_shard_grid_inner_dim());
+
+        std::size_t worker_y_offset = (!tensor_shard_spec.transposed_grid * shard_grid_outer_dim_index) + (tensor_shard_spec.transposed_grid * shard_grid_inner_dim_index);
+        std::size_t worker_x_offset = (!tensor_shard_spec.transposed_grid * shard_grid_inner_dim_index) + (tensor_shard_spec.transposed_grid * shard_grid_outer_dim_index);
+
+        std::size_t page_in_shard_x = page_global_inner_dim - (global_shard_index * tensor_shard_spec.get_pages_per_shard_x());
+        std::size_t page_in_shard_y = page_global_outer_dim;
+        std::size_t page_offset_in_shard = (page_global_outer_dim * tensor_shard_spec.get_pages_per_shard_x()) + page_in_shard_x;
+
+        std::size_t worker_x_logical = tensor_shard_spec.shard_grid_start_x_logical + worker_x_offset;
+        std::size_t worker_y_logical = tensor_shard_spec.shard_grid_start_y_logical + worker_y_offset;
+
+        noc_grid_index_t noc_x = worker_to_noc_lookup.get_noc_x_from_worker_x(worker_x_logical);
+        noc_grid_index_t noc_y = worker_to_noc_lookup.get_noc_y_from_worker_y(worker_y_logical);
+
+        return test_shard_location_t{device_core_location_t{noc_y, noc_x}, page_offset_in_shard};
+    }
+
+    // uint64_t get_noc_addr(std::size_t global_page_id, std::size_t offset = 0) const {
```

Comment:
If we can stub out get_noc_addr so that if building on host, we don't include noc async calls, then we can enable this and also remove all of the macros that check for mem-layout and data layout from within the all the ccl packetize/depacketize helpers.

i.e. the code paths that look like this will all be removed and we'll have true single code path for any combo of {row-major vs tiled}x{sharded vs interleaved vs single bank}
```
#elif defined TILED_LAYOUT
    #ifdef INTERLEAVED_MEM_LAYOUT
        noc_async_write_tile(curr_page_idx, d, l1_read_addr);
    #elif defined SHARDED_MEM_LAYOUT
```


(alternatively, a simple macro that tells us if we are compiling for device will let us provide alternative implementations.